pkg/raft/rawnode.go

```diff
@@ -122,80 +122,70 @@ func (rn *RawNode) Step(m pb.Message) error {
 		rd.HardState = hardSt
+		rn.prevHardSt = hardSt
 	}
+
 	if r.raftLog.hasNextUnstableSnapshot() {
 		rd.Snapshot = *r.raftLog.nextUnstableSnapshot()
 	}
-	rd.MustSync = MustSync(r.hardState(), rn.prevHardSt, len(rd.Entries))
+	if r.raftLog.hasNextUnstableEnts() {
+		rd.Entries = r.raftLog.nextUnstableEnts()
+	}
+	// TODO(pav-kv): remove "accept" methods down the stack, since we now accept
+	// all updates unconditionally.
+	r.raftLog.acceptUnstable()
+	rd.MustSync = MustSync(hardSt, prevHardSt, len(rd.Entries))
+
+	allowUnstable := rn.applyUnstableEntries()
+	if r.raftLog.hasNextCommittedEnts(allowUnstable) {
+		entries := r.raftLog.nextCommittedEnts(allowUnstable)
+		index := entries[len(entries)-1].Index
+		r.raftLog.acceptApplying(index, entsSize(entries), allowUnstable)
+		rd.CommittedEntries = entries
+	}
 
 	if rn.asyncStorageWrites {
-		// If async storage writes are enabled, enqueue messages to
-		// local storage threads, where applicable.
+		// If async storage writes are enabled, enqueue messages to local storage
+		// threads, where applicable.
 		if needStorageAppendMsg(r, rd) {
-			m := newStorageAppendMsg(r, rd)
-			rd.Messages = append(rd.Messages, m)
+			rd.Messages = append(rd.Messages, newStorageAppendMsg(r, rd))
 		}
 		if needStorageApplyMsg(rd) {
-			m := newStorageApplyMsg(r, rd)
-			rd.Messages = append(rd.Messages, m)
+			rd.Messages = append(rd.Messages, newStorageApplyMsg(r, rd))
 		}
 	} else {
-		// If async storage writes are disabled, immediately enqueue
-		// msgsAfterAppend to be sent out. The Ready struct contract
-		// mandates that Messages cannot be sent until after Entries
-		// are written to stable storage.
+		// If async storage writes are disabled, immediately enqueue msgsAfterAppend
+		// to be sent out. The Ready struct contract mandates that Messages cannot
+		// be sent until after Entries are written to stable storage.
 		for _, m := range r.msgsAfterAppend {
 			if m.To != r.id {
 				rd.Messages = append(rd.Messages, m)
-			}
-		}
-	}
-	// Accept the Ready now.
-
-	if rd.SoftState != nil {
-		rn.prevSoftSt = rd.SoftState
-	}
-	if !IsEmptyHardState(rd.HardState) {
-		rn.prevHardSt = rd.HardState
-	}
-	if !rn.asyncStorageWrites {
-		if len(rn.stepsOnAdvance) != 0 {
-			r.logger.Panicf("two accepted Ready structs without call to Advance")
-		}
-		for _, m := range r.msgsAfterAppend {
-			if m.To == r.id {
+			} else {
 				rn.stepsOnAdvance = append(rn.stepsOnAdvance, m)
 			}
 		}
 		if needStorageAppendRespMsg(rd) {
-			m := newStorageAppendRespMsg(r, rd)
-			rn.stepsOnAdvance = append(rn.stepsOnAdvance, m)
+			rn.stepsOnAdvance = append(rn.stepsOnAdvance,
+				newStorageAppendRespMsg(r, rd))
 		}
 		if needStorageApplyRespMsg(rd) {
-			m := newStorageApplyRespMsg(r, rd.CommittedEntries)
-			rn.stepsOnAdvance = append(rn.stepsOnAdvance, m)
+			rn.stepsOnAdvance = append(rn.stepsOnAdvance,
+				newStorageApplyRespMsg(r, rd.CommittedEntries))
 		}
 	}
-	r.msgs = nil
 	r.msgsAfterAppend = nil
```

Comment:
Need to double-check that this field is handled equivalently. Previously it was cleared between `readyWithoutAccept` and `acceptReady`, now (in the merged code) it is cleared at the end.